### ISSUE TITLE
adding multiple block volumes to one VM fails with error

### ISSUE THREAD

**Issue by FettBoba** [oracle/terraform-provider-baremetal#73]
After attaching one block volume to a vm running terraform plan again results in the following warning that indicated that the attachment would be destroyed on second apply
 
 
-/+ baremetal_core_volume_attachment.Web1Block0Attach
    attachment_type:     "" => "iscsi" (forces new resource)
    availability_domain: "YZAF:PHX-AD-1" => "<computed>"
    compartment_id:     
 …
 
So there are two issues:
1 Once one block volume is attached, the attachment is destroyed and re-created on repeated apply runs
(As a consequence no additional block volumes can be attached). – see error message below
 
2 Trying to apply two block volumes on the first run of course is not possible either – only one attach can run at one time. – see error message below
 
error message:
 
  …
baremetal_core_volume_attachment.Web1Block0Attach: Destroying...
baremetal_core_volume.Web1Block1: Creating...
  …
baremetal_core_volume.Web1Block1: Creation complete
baremetal_core_volume_attachment.Web1Block1Attach: Creating...
baremetal_core_volume_attachment.Web1Block0Attach: Still destroying... (10s elapsed)
baremetal_core_volume_attachment.Web1Block0Attach: Still destroying... (20s elapsed)
baremetal_core_volume_attachment.Web1Block0Attach: Destruction complete
baremetal_core_volume_attachment.Web1Block0Attach: Creating...
baremetal_core_volume_attachment.Web1Block0Attach: Still creating... (3s elapsed)
baremetal_core_volume_attachment.Web1Block0Attach: Still creating... (10s elapsed)
baremetal_core_volume_attachment.Web1Block0Attach: Still creating... (13s elapsed)
baremetal_core_volume_attachment.Web1Block0Attach: Creation complete
Error applying plan:
1 error(s) occurred:
* baremetal_core_volume_attachment.Web1Block1Attach: Code: Conflict; OPC Request ID: /DA9BE15C20AC40799253C63F35D63E9F/2E8A98331CB74375BD229E54A907E374; Message: instance ocid1.instance.oc1.phx.abyhqljsmhblbxcyefx4mv7mv66jysfmzccbpn2iqc7s4odubkzn7d6b62gq is currently being modified, try again later

**Comment by craigcarl-oracle** [comment]
@mikejihbe @josephholsten the lifecycle, ignore_changes workaround won't work here, the number of attached volumes is mutable. I P0'ed it, the unexpected detach of a block volume can lead to data loss.

**Comment by mikejihbe** [comment]
I cannot reproduce the volume destruction (data loss) issue. Please share a fully working minimal example.

Here's mine tf file. This does reproduce the multiple volume conflict error. We don't loop on "try again later" issues, but we probably should. 
```
provider "baremetal" {
  tenancy_ocid = "${var.tenancy_ocid}"
  user_ocid = "${var.user_ocid}"
  fingerprint = "${var.fingerprint}"
  private_key_path = "${var.private_key_path}"
}

variable "AD" {
    default = "1"
}

variable "256GB" {
    default = "262144"
}

variable "InstanceShape" {
    default = "VM.Standard1.2"
}

variable "InstanceOS" {
    default = "Oracle Linux"
}

variable "InstanceOSVersion" {
    default = "7.3"
}
variable "VPC-CIDR" {
  default = "10.0.0.0/16"
}

resource "baremetal_identity_compartment" "test-dec5" {
  name = "test_compartment_dec5"
  description = "A special test compartment."
}

resource "baremetal_core_virtual_network" "CompleteVCN" {
  cidr_block = "${var.VPC-CIDR}"
  compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
  display_name = "CompleteVCN"
}

resource "baremetal_core_internet_gateway" "CompleteIG" {
    compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
    display_name = "CompleteIG"
    vcn_id = "${baremetal_core_virtual_network.CompleteVCN.id}"
}

resource "baremetal_core_route_table" "RouteForComplete" {
    compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
    vcn_id = "${baremetal_core_virtual_network.CompleteVCN.id}"
    display_name = "RouteTableForComplete"
    route_rules {
        cidr_block = "0.0.0.0/0"
        network_entity_id = "${baremetal_core_internet_gateway.CompleteIG.id}"
    }
}

resource "baremetal_core_subnet" "WebSubnetAD1" {
  availability_domain = "${lookup(data.baremetal_identity_availability_domains.ADs.availability_domains[0],"name")}"
  cidr_block = "10.0.1.0/24"
  display_name = "WebSubnetAD1"
  compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
  vcn_id = "${baremetal_core_virtual_network.CompleteVCN.id}"
  route_table_id = "${baremetal_core_route_table.RouteForComplete.id}"
  security_list_ids = ["${baremetal_core_security_list.WebSubnet.id}"]
}


resource "baremetal_core_security_list" "WebSubnet" {
    compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
    display_name = "Public"
    vcn_id = "${baremetal_core_virtual_network.CompleteVCN.id}"
    egress_security_rules = [{
        destination = "0.0.0.0/0"
        protocol = "6"
    }]
    ingress_security_rules = [{
        tcp_options {
            "max" = 80
            "min" = 80
        }
        protocol = "6"
        source = "0.0.0.0/0"
    },
	{
	protocol = "6"
	source = "${var.VPC-CIDR}"
    }]
}

data "baremetal_identity_availability_domains" "ADs" {
  compartment_id = "${var.tenancy_ocid}"
}

data "baremetal_core_images" "OLImageOCID" {
    compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
    operating_system = "${var.InstanceOS}"
    operating_system_version = "${var.InstanceOSVersion}"
}

resource "baremetal_core_instance" "TFInstance" {
  availability_domain = "${lookup(data.baremetal_identity_availability_domains.ADs.availability_domains[var.AD - 1],"name")}"
  compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
  display_name = "TFInstance"
  image = "${lookup(data.baremetal_core_images.OLImageOCID.images[0], "id")}"
  shape = "${var.InstanceShape}"
  subnet_id = "${baremetal_core_subnet.WebSubnetAD1.id}"
  metadata {
    ssh_authorized_keys = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC+UC9MFNA55NIVtKPIBCNw7++ACXhD0hx+Zyj25JfHykjz/QU3Q5FAU3DxDbVXyubgXfb/GJnrKRY8O4QDdvnZZRvQFFEOaApThAmCAM5MuFUIHdFvlqP+0W+ZQnmtDhwVe2NCfcmOrMuaPEgOKO3DOW6I/qOOdO691Xe2S9NgT9HhN0ZfFtEODVgvYulgXuCCXsJs+NUqcHAOxxFUmwkbPvYi0P0e2DT8JKeiOOC8VKUEgvVx+GKmqasm+Y6zHFW7vv3g2GstE1aRs3mttHRoC/JPM86PRyIxeWXEMzyG5wHqUu4XZpDbnWNxi6ugxnAGiL3CrIFdCgRNgHz5qS1l Mike@MIHBE-M-RCNE"
    user_data = "${base64encode(file("./docs/examples/compute/single-instance/userdata/bootstrap"))}"
  }

  timeouts {
    create = "60m"
  }
}

resource "baremetal_core_volume" "TFBlock0" {
  availability_domain = "${lookup(data.baremetal_identity_availability_domains.ADs.availability_domains[var.AD - 1],"name")}"
  compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
  display_name = "TFBlock0"
  size_in_mbs = "${var.256GB}"
}

resource "baremetal_core_volume_attachment" "TFBlock0Attach" {
    attachment_type = "iscsi"
    compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
    instance_id = "${baremetal_core_instance.TFInstance.id}"
    volume_id = "${baremetal_core_volume.TFBlock0.id}"
}

resource "baremetal_core_volume" "TFBlock1" {
  availability_domain = "${lookup(data.baremetal_identity_availability_domains.ADs.availability_domains[var.AD - 1],"name")}"
  compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
  display_name = "TFBlock1"
  size_in_mbs = "${var.256GB}"
}

resource "baremetal_core_volume_attachment" "TFBlock1Attach" {
    attachment_type = "iscsi"
    compartment_id = "${baremetal_identity_compartment.test-dec5.id}"
    instance_id = "${baremetal_core_instance.TFInstance.id}"
    volume_id = "${baremetal_core_volume.TFBlock1.id}"
}

```

**Comment by FettBoba** [comment]
My config consists of a number of separate files, so it'll take some time to create a minimal example.

But I guess the problem can of course be reproduced in your example.

After first apply run and attachment just for this test sake I ran only manually:

# sudo iscsiadm -m node -o new -T iqn.2015-12.com.oracleiaas:809f6b11-aa3b-426c-ba31-dc334fe96a46 -p 169.254.0.2:3260
# sudo iscsiadm -m node -o update -T iqn.2015-12.com.oracleiaas:809f6b11-aa3b-426c-ba31-dc334fe96a46 -n node.startup -v automatic
# sudo iscsiadm -m node -T iqn.2015-12.com.oracleiaas:809f6b11-aa3b-426c-ba31-dc334fe96a46 -p 169.254.0.2:3260 -l
# lsblk

NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sda      8:0    0 46.6G  0 disk 
├─sda1   8:1    0  544M  0 part /boot/efi
├─sda2   8:2    0    8G  0 part [SWAP]
└─sda3   8:3    0   38G  0 part /
sdb      8:16   0  256G  0 disk
  

# sudo mkfs.xfs /dev/sdb
# sudo mkdir /blockstor
# sudo mount -t xfs /dev/sdb /blockstor
# sudo chown opc:opc /blockstor/
# echo date >> /blockstor/Testfile.txt

On every re-run of of terraform apply the volume gets detached and attached.

So afterwards the mount is somehow gone:
# ls -l /blockstor/
ls: cannot access /blockstor/: Input/output error
# cat /blockstor/Testfile.txt
cat: /blockstor/Testfile.txt: Input/output error

although df tells me:

# df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda3        38G  1.7G   34G   5% /
tmpfs           6.8G     0  6.8G   0% /dev/shm
/dev/sda1       543M  280K  543M   1% /boot/efi
/dev/sdb        256G   33M  256G   1% /blockstor

and mounting is also not possible:
# sudo mount -t xfs /dev/sdb /blockstor 
mount: /dev/sdb already mounted or /blockstor busy
mount: according to mtab, /dev/sdb is already mounted on /blockstor

What helps is:

# sudo umount -f /blockstor 
# sudo mount -t xfs /dev/sdb /blockstor

File is still there and readable, but it was unusable until remount...not nice either ;-)

But my original issue was that I can **never** attach more than one block volume: 
1. not initially, because only one attach can run at a time
2. not in sequential runs, because every time the block volume is detached&re-attached which leads to situation 1

 - is there a workaround?

**Comment by craigcarl-oracle** [comment]
@mikejihbe @josephholsten Take a look at the volume attach behavior, we need to be paying attention to the lifecycle state of each attach and not attempting second attach until the first has completed. It looks like we're trying multiple attaches in parallel and retrying on error.

**Comment by mikejihbe** [comment]
This seems like 2 things, so I want to make sure I understand.

### This volume attachment issue Fett is seeing

I still can't get a `terraform plan` to say it needs to reattach the volume. I've confirmed that a second apply only performs GET requests against the API. If the volume is being detached, the TF provider is not doing it.

### Multiple Volume Attach behavior

You're right. We pause and retry on "409 Conflict - try again later" responses from the API. This does happen when multiple attachments happen in parallel. It's literally what the API says to do, and I think it actually makes sense as the expected behavior.

I think you're saying you'd like them to run exclusively serially so this situation doesn't happen?

This is theoretically possible but is not really the "Terraform Way." The provider doesn't know anything about dependencies between objects. Terraform does. We could add serialization here, but it would be a LOT more complicated.

There is a workaround to do this. One could have `depends_on` links chained between one's volume attachments.

Thoughts?

**Comment by FettBoba** [comment]
**This volume attachment issue Fett is seeing**
It's not shown in terraform plan - it just fails on apply with the posted error. Only in the Cloud GUI you see the block volume getting detached, then it's gone for a few seconds and shows up as reattaching.
Tht's why it was surprising that we could constaltly r/w to it without any noticeable interruption.

Multiple Volume Attach behavior
If depends on is the only solution/workaround I can solve our issue. It's just not pretty if the attach of any 1+n volume simply fails in case no depend on logic is applied.

**Comment by rcohenma** [comment]
This issue was fixed in v1.0.14
